packages/collapsible-side-nav/src/SideNav.tsx

```diff
@@ -0,0 +1,204 @@
+    box-shadow ${transitionDurationMilliseconds}ms ease-in-out;
+`;
+
+const collapsedNavStyle = css`
+  position: absolute;
+  min-width: 48px;
+  width: 48px;
+  top: 0;
+  bottom: 0;
+`;
+
+const collapsedHoveredNavStyle = css`
+  position: absolute;
+  top: 0;
+  bottom: 0;
+  border-width: 0;
+  box-shadow: 6px 0 10px -8px ${transparentize(0.7, uiColors.black)};
+`;
+
+const collapsedSpacerStyle = css`
+  min-width: 48px;
+  width: 48px;
+`;
+
+const navListStyle = css`
+  position: relative;
+  padding-top: 12px;
+  height: 100%;
+  overflow: hidden;
+  transition: all ${transitionDurationMilliseconds}ms ease-in-out;
+`;
+
+const collapsedNavListStyle = css`
+  padding-top: 64px;
+`;
+
+type Props = {
+  children: React.ReactNode;
+  className?: string;
+  currentPath?: string;
+} & (
+  | { collapsible: false; initialCollapsed?: undefined }
+  | { collapsible?: true; initialCollapsed?: boolean }
+) &
+  Omit<HTMLElementProps<'nav'>, 'ref'>;
+
+const idAllocator = IdAllocator.create('side-nav');
+
+const SideNav = React.forwardRef<HTMLElement, Props>(function SideNav(
+  {
+    children,
+    className,
+    currentPath,
+    collapsible = true,
+    initialCollapsed = false,
+    ...rest
+  }: Props,
+  ref,
+) {
+  const id = useMemo(() => idAllocator.generate(), []);
+
+  const [collapsed, setCollapsed] = useState(initialCollapsed);
+
+  // Manually detect hover so that hover state can be passed through the context.
+  // Using pure CSS would require the nav component to import a data prop selector
+  // from every sub-component or an equivalent amount of work.
+  const [hovered, setHovered] = useState(false);
+
+  // Debounce hover changes to prevent jitter when rapidly moving the mouse
+  // on and off the nav. Alternatively we could set transition-delay for this
+  // and all sub-components, but that would be less convenient and wouldn't
+  // prevent unnecessary re-renders.
+  const debouncedSetHovered = useMemo(
+    () => ({
+      toTrue: debounce(() => setHovered(true), transitionDurationMilliseconds),
+      toFalse: debounce(
+        () => setHovered(false),
+        transitionDurationMilliseconds,
+      ),
+    }),
+    [],
+  );
+
+  const onMouseOver = useCallback(() => {
+    debouncedSetHovered.toFalse.cancel();
+    debouncedSetHovered.toTrue();
+  }, [debouncedSetHovered]);
+
+  const onMouseLeave = useCallback(() => {
+    debouncedSetHovered.toTrue.cancel();
+    debouncedSetHovered.toFalse();
+  }, [debouncedSetHovered]);
+
+  useEffect(() => {
+    if (!collapsible) {
+      debouncedSetHovered.toTrue.cancel();
+      setHovered(false);
+
+      setCollapsed(false);
+    }
+  }, [collapsible, debouncedSetHovered]);
+
+  const toggleCollapse = useCallback(() => {
+    debouncedSetHovered.toTrue.cancel();
+    setHovered(false);
+
+    setCollapsed(collapsed => !collapsed);
+  }, [debouncedSetHovered]);
+
+  useEventListener(
+    'keydown',
+    event => {
+      if (event.keyCode === 219 /* BracketLeft */) {
```

Comment:
Lets add to our keyMap in lib (it seems that keyCode is starting to be deprecated anyway, so when we change how we approach key-detection we'll want to make sure we do this everywhere at once)